test/sys/test_pthread.rs

```diff
@@ -20,3 +20,103 @@ fn test_pthread_kill_none() {
     pthread_kill(pthread_self(), None)
         .expect("Should be able to send signal to my thread.");
 }
+
+#[test]
+#[cfg(target_os = "linux")]
+fn test_pthread_mutex_wrapper() {
+    use nix::{
+        sys::{
+            mman::{mmap, MapFlags, ProtFlags},
+            pthread::{Mutex, MutexAttr},
+        },
+        unistd::{fork, ForkResult},
+    };
+    use std::{mem::size_of, num::NonZeroUsize, os::unix::io::OwnedFd};
+    struct MutexWrapper {
+        lock: Mutex,
+        data: u128,
+    }
+    impl MutexWrapper {
+        fn add(&mut self) {
+            self.lock.lock().unwrap();
+            self.data += 1;
+            self.lock.unlock().unwrap();
+        }
+    }
+
+    /// Number of forks to spawn that mutate the data, 2^n processes.
+    const FORKS: usize = 3;
+    /// Number of threads each process spawns that mutate the data.
+    const THREADS: usize = 20;
+    /// Number of iterations each thread mutates the data.
+    const ITERATIONS: usize = 100_000;
+
+    let mut mutex_attr = MutexAttr::new().unwrap();
+    mutex_attr.set_shared(true).unwrap();
+
+    let mutex_wrapper = unsafe {
+        mmap::<OwnedFd>(
+            None,
+            NonZeroUsize::new_unchecked(size_of::<MutexWrapper>()),
+            ProtFlags::PROT_WRITE | ProtFlags::PROT_READ,
+            MapFlags::MAP_SHARED | MapFlags::MAP_ANONYMOUS,
+            None,
+            0,
+        )
+        .unwrap()
+        .cast()
+    };
+
+    unsafe {
+        std::ptr::write(
+            mutex_wrapper,
+            MutexWrapper {
+                lock: Mutex::new(Some(mutex_attr)).unwrap(),
+                data: 0,
+            },
+        );
+    }
+
+    let fork_results = (0..FORKS)
+        .map(|_| unsafe { fork().unwrap() })
+        .collect::<Vec<_>>();
+
+    let handles = (0..THREADS)
+        .map(|_| {
+            let bits = mutex_wrapper as usize;
+            std::thread::spawn(move || {
+                let ptr = bits as *mut MutexWrapper;
```

Comment:
Here you create a mutable pointer from a shared reference, and it looks like you had to cast that shared reference to a `usize` just to fool the compiler into allowing it.  Dereferencing the shared reference on line 90 is UB.  This is exactly the kind of pattern that Mutexes should be used to prevent.

Reply:
I've changed it to pass a reference.